File: src/main/java/game/Rei.java
package game;

public class Rei extends Peca {
    public Rei(String cor) {
        super(cor);
        this.nomePeca = "Rei";
    }

    @Override
    public boolean isMovimentoValido(Tabuleiro tabuleiro, int linhaInicial, int colunaInicial, int linhaFinal, int colunaFinal) {
        
        int diffLinha = Math.abs(linhaInicial - linhaFinal);
        int diffColuna = Math.abs(colunaInicial - colunaFinal);
        boolean podeMover = diffLinha <= 1 && diffColuna <= 1;

        if (podeMover) {
            Peca pecaDestino = tabuleiro.getPeca(linhaFinal, colunaFinal);
            // Pode mover se a casa de destino estiver vazia OU tiver uma peça de cor diferente
            if (pecaDestino == null || !pecaDestino.getCor().equals(this.getCor())) {
                return true;
            }
        }

        return false; // Movimento inválido
    }

    @Override
    public String toString() {
        return "RE{}";
    }
}

File: Projeto JS/src/game/Rei.js
import { Peca } from './Peca.js';

export class Rei extends Peca {
    constructor(cor) {
        super(cor);
        this.nomePeca = "Rei";
    }

    isMovimentoValido(tabuleiro, linhaInicial, colunaInicial, linhaFinal, colunaFinal) {
        const pecaDestino = tabuleiro.getPeca(linhaFinal, colunaFinal);

        if (pecaDestino !== null && pecaDestino.getCor() === this.getCor()) {
            return false;
        }

        const diffLinha = Math.abs(linhaFinal - linhaInicial);
        const diffColuna = Math.abs(colunaFinal - colunaInicial);

        return diffLinha <= 1 && diffColuna <= 1;
    }

    toString() {
        return "R{}";
    }
}